src/imports/background/import-item-creation.js

```diff
@@ -0,0 +1,195 @@
+import { differMaps } from 'src/util/map-set-helpers'
+import {
+    IMPORT_TYPE,
+    OLD_EXT_KEYS,
+    DEF_ALLOW,
+} from 'src/options/imports/constants'
+
+// Binds old ext bookmark urls to a function that transforms old ext data to an import item
+const transformOldExtDataToImportItem = bookmarkUrls => (item, index) => ({
+    type: IMPORT_TYPE.OLD,
+    url: item.url,
+    timestamp: item.time,
+    hasBookmark: bookmarkUrls.has(item.url),
+    index,
+})
+
+// Binds an import type to a function that transforms a history/bookmark doc to an import item.
+const transformBrowserItemToImportItem = type => item => ({
+    browserId: item.id,
+    url: item.url,
+    // HistoryItems contain lastVisitTime while BookmarkTreeNodes contain dateAdded
+    timestamp: item.lastVisitTime || item.dateAdded,
+    type,
+})
+
+async function checkWithExistingBookmarks() {
+    const { rows: existingBmDocs } = await db.allDocs({
+        startkey: bookmarkKeyPrefix,
+        endkey: `${bookmarkKeyPrefix}\uffff`,
+    })
+
+    const existingIds = new Set(
+        existingBmDocs.map(row => convertBookmarkDocId(row.id).url),
+    )
+    return url => existingIds.has(url)
+}
+
+/**
+ * @returns {({ url: string }) => boolean} A function affording checking of a URL against the
+ *  URLs of existing page docs.
+ */
+async function checkWithExistingPages() {
+    const { rows: existingPageDocs } = await db.allDocs({
+        startkey: pageKeyPrefix,
+        endkey: `${pageKeyPrefix}\uffff`,
+    })
+
+    const existingIds = new Set(
+        existingPageDocs.map(row => convertPageDocId(row.id).url),
+    )
+    return url => existingIds.has(url)
+}
+
+async function initFilterItemsByUrl() {
+    const isNotBlacklisted = await checkWithBlacklist()
+
+    /**
+     * Performs all needed filtering on a collection of history, bookmark, or old ext items. As these
+     * can be quite large, attemps to do in a single iteration of the input collection.
+     *
+     * @param {Array<any>} items Array of items with `url` to filter on.
+     * @param {(item: any) => any} transform Opt. transformformation fn turning current iterm into import item structure.
+     * @param {(url: string) => bool} alreadyExists Opt. checker function to check against existing data.
+     * @return {Map<string, any>} Filtered version of `items` array made into a Map of encoded URL strings to import items.
+     */
+    return (items, transform = f => f, alreadyExists) =>
+        items.reduce((accItems, currItem) => {
+            if (!isLoggable(currItem) || !isNotBlacklisted(currItem)) {
+                return accItems
+            }
+
+            try {
+                // Augment the item with encoded URL for DB check and item map
+                const encodedUrl = encodeUrl(currItem.url, false)
+
+                // Filter out items already existing in DB, if specified
+                if (alreadyExists == null || !alreadyExists(encodedUrl)) {
```

Comment:
the alreadyExists check can be probably move on line 76, non need to enter the try catch and encode the url for it

Reply:
`encodeUrl` should throw a malformed URI error for certain URLs that have placeholders in them. Like `%s` in `http://www.youtube.com/results?search_query=%s` commonly used in keyword search bookmarks), hence the `try...catch` here.

Reply:
I understand, but you can reduce the try catch scope to encodeUrl only instead of the entire block